readme.adoc
:response-version: 0.1.0
:javac-version: 0.0.1
:fsm-version: 0.0.1-SNAPSHOT
:schema-version: 0.0.1-SNAPSHOT

image:https://img.shields.io/github/license/just-sultanov/ninja.platform[license,link=license]
image:https://codecov.io/gh/just-sultanov/ninja.platform/branch/master/graph/badge.svg?token=HVEZAXE27E)[codecov,link=https://codecov.io/gh/just-sultanov/ninja.platform]
image:https://github.com/just-sultanov/ninja.platform/workflows/build/badge.svg[build]

== ninja.platform

Build up projects like a ninja.
This repository includes useful libraries (aka ninja tools) for everyday use.

image:docs/design/logotype.full.transparent.png[logotype,800]

=== ninja tools

[%autowidth]
|===
| # | Module | Description | Version | Badges

| {counter:module}
| link:docs/ninja.tools/response.adoc[ninja.platform/response]
| A Clojure(Script) library for the unified responses
| `{response-version}`
| image:https://codecov.io/gh/just-sultanov/ninja.platform/branch/master/graph/badge.svg?token=HVEZAXE27E&flag=response[codecov,link=https://codecov.io/gh/just-sultanov/ninja.platform]
image:https://cljdoc.org/badge/ninja.platform/response[cljdoc,link=https://cljdoc.org/d/ninja.platform/response/CURRENT]
image:https://img.shields.io/clojars/v/ninja.platform/response.svg[clojars,link=https://clojars.org/ninja.platform/response]

| {counter:module}
| link:docs/ninja.tools/javac.adoc[ninja.platform/javac]
| A Clojure library for compiling Java source code
| `{javac-version}`
| image:https://codecov.io/gh/just-sultanov/ninja.platform/branch/master/graph/badge.svg?token=HVEZAXE27E&flag=javac[codecov,link=https://codecov.io/gh/just-sultanov/ninja.platform]
image:https://cljdoc.org/badge/ninja.platform/javac[cljdoc,link=https://cljdoc.org/d/ninja.platform/javac/CURRENT]
image:https://img.shields.io/clojars/v/ninja.platform/javac.svg[clojars,link=https://clojars.org/ninja.platform/javac]

//| {counter:module}
//| link:docs/ninja.tools/fsm.adoc[ninja.platform/fsm]
//| A Clojure(Script) library for the finite state machines
//| `{fsm-version}`
//| image:https://codecov.io/gh/just-sultanov/ninja.platform/branch/master/graph/badge.svg?token=HVEZAXE27E&flag=fsm[codecov,link=https://codecov.io/gh/just-sultanov/ninja.platform]
//image:https://cljdoc.org/badge/ninja.platform/fsm[cljdoc,link=https://cljdoc.org/d/ninja.platform/fsm/CURRENT]
//image:https://img.shields.io/clojars/v/ninja.platform/fsm.svg[clojars,link=https://clojars.org/ninja.platform/fsm]

//| {counter:module}
//| link:docs/ninja.tools/schema.adoc[ninja.platform/schema]
//| A Clojure(Script) library for the data schemas
//| `{schema-version}`
//| image:https://codecov.io/gh/just-sultanov/ninja.platform/branch/master/graph/badge.svg?token=HVEZAXE27E&flag=schema[codecov,link=https://codecov.io/gh/just-sultanov/ninja.platform]
//image:https://cljdoc.org/badge/ninja.platform/schema[cljdoc,link=https://cljdoc.org/d/ninja.platform/schema/CURRENT]
//image:https://img.shields.io/clojars/v/ninja.platform/schema.svg[clojars,link=https://clojars.org/ninja.platform/schema]

|===

=== License

Copyright © 2021 Ilshat Sultanov.

Distributed under the Eclipse Public License version 1.0.


changelog.adoc
== Changelog

=== ninja tools

- xref:response[ninja.platform/response]
- xref:javac[ninja.platform/javac]

[[response]]
=== ninja.platform/response

==== Released

===== v0.1.0 (2021-02-05)

====== Added

A very simple API for the unified responses.


Anomalies registry::
- `ninja.response/add-anomaly!` - Adds a new anomaly to the anomalies registry
- `ninja.response/remove-anomaly!` - Removes an anomaly from the anomalies registry
- `ninja.response/anomaly?` - Returns `true` if `x` is an anomaly.
Otherwise, `false`


Predefined anomalies::
- `:error`
- `:warning`
- `:exception`
- `:unavailable`
- `:interrupted`
- `:incorrect`
- `:unauthorized`
- `:forbidden`
- `:not-found`
- `:unsupported`
- `:conflict`
- `:busy`
- `:unknown`


Unified response::
- `ninja.response/IResponse` - A unified response protocol
- `ninja.response/Response` - A default implementation of the unified response protocol


Unified response builder::
- `ninja.response/as-response` - Returns an instance of a unified response


Unified error response helpers::
- `ninja.response/as-error` - Returns a unified `error` response
- `ninja.response/as-warning` - Returns a unified `warning` response
- `ninja.response/as-exception` - Returns a unified `exception` response
- `ninja.response/as-unavailable` - Returns a unified `unavailable` response
- `ninja.response/as-interrupted` - Returns a unified `interrupted` response
- `ninja.response/as-incorrect` - Returns a unified `incorrect` response
- `ninja.response/as-unauthorized` - Returns a unified `unauthorized` response
- `ninja.response/as-forbidden` - Returns a unified `forbidden` response
- `ninja.response/as-not-found` - Returns a unified `not-found` response
- `ninja.response/as-unsupported` - Returns a unified `unsupported` response
- `ninja.response/as-conflict` - Returns a unified `conflict` response
- `ninja.response/as-busy` - Returns a unified `busy` response
- `ninja.response/as-unknown` - Returns a unified `unknown` response


Unified success response helpers::
- `ninja.response/as-success` - Returns a unified `success` response
- `ninja.response/as-found` - Returns a unified `found` response
- `ninja.response/as-created` - Returns a unified `created` response
- `ninja.response/as-updated` - Returns a unified `updated` response
- `ninja.response/as-deleted` - Returns a unified `deleted` response
- `ninja.response/as-accepted` - Returns a unified `accepted` response


Helper macros::
- `ninja.response/safe` - Extended version of try-catch


[[javac]]
=== ninja.platform/javac

==== Released

===== v0.0.1 (2021-03-16)

====== Added

A very simple API for compiling Java source code.


javac/src/test/clojure/ninja/javac_test.clj
(ns ninja.javac-test
  (:require
    [clojure.string :as string]
    [clojure.test :refer [deftest testing is]]
    [ninja.javac :as sut]
    [ninja.response :as r])
  (:import
    (java.nio.file
      Path)))


(deftest make-path-test
  (testing "build path"
    (is (= "a" (str (sut/make-path "a"))))
    (is (= "a/b/c" (str (sut/make-path "a/b" "c")) (str (sut/make-path "a" "b" "c"))))))


(deftest make-classpath-test
  (testing "calculate classpath"
    (let [includes? #(string/includes? % "ninja/platform/schema/0.0.1-alpha1/schema-0.0.1-alpha1.jar")]
      (testing "without any arguments"
        (is (false? (includes? (sut/make-classpath))))
        (is (false? (includes? (sut/make-classpath {:deps-map nil}))))
        (is (false? (includes? (sut/make-classpath {:deps-map {}})))))

      (testing "with provided aliases"
        (is (false? (includes? (sut/make-classpath {:deps-map {}, :aliases [:module.test/deps]}))))
        (is (true? (includes? (sut/make-classpath {:deps-map nil, :aliases [:module.test/deps]}))))
        (is (true? (includes? (sut/make-classpath {:aliases [:module.test/deps]}))))))))


(deftest as-path-test
  (testing "expected an instance of `java.nio.file.Path`"
    (is (true? (instance? Path (sut/as-path "a/b/c"))))
    (is (true? (instance? Path (sut/as-path (sut/make-path "a/b/c")))))))


(deftest find-paths-test
  (testing "find paths"
    (testing "in the non-existing directory"
      (is (zero? (count (sut/find-paths "fake/path")))))
    (testing "in the fixtures directory"
      (is (= 9 (count (sut/find-paths "src/test/resources/fixtures")))))))


(deftest find-java-paths-test
  (testing "find paths to Java source files"
    (testing "in the non-existing directory"
      (is (zero? (count (sut/find-java-paths "fake/path")))))
    (testing "in the fixtures directory"
      (is (= 8 (count (sut/find-java-paths "src/test/resources/fixtures")))))))


(deftest create-delete-dirs-test
  (testing "directories creation and deletion"
    (let [root-path   "a"
          nested-path (sut/make-path root-path "b/c/d")]
      (testing "directory shouldn't be exists"
        (is (false? (sut/exists? nested-path))))
      (testing "directory should be created"
        (sut/create-dirs! nested-path)
        (is (true? (sut/exists? nested-path))))
      (testing "nested directory should be deleted"
        (sut/delete-dirs! nested-path)
        (is (false? (sut/exists? nested-path))))
      (testing "root directory should be deleted"
        (sut/delete-dirs! root-path)
        (is (false? (sut/exists? root-path)))))))


(deftest make-command-test
  (testing "build compiler command"
    (testing "with default options"
      (is (= ["-cp" "some:classpath" "-d" "classes" "fake/Example.java"]
            (sut/make-command {:classpath    "some:classpath"
                               :source-paths ["fake/Example.java"]}))))
    (testing "with all options"
      (is (= ["-cp" "some:classpath" "-target" "15" "-source" "15" "-Xlint:-options" "-d" "target/classes" "fake/Example.java"]
            (sut/make-command {:classpath        "some:classpath"
                               :target-path      "target/classes"
                               :compiler-options ["-target" "15"
                                                  "-source" "15"
                                                  "-Xlint:-options"]
                               :source-paths     ["fake/Example.java"]}))))))


(deftest compile-test
  (let [source-path      "src/test/resources/fixtures"
        target-path      (str "target/classes/" (System/nanoTime))
        compiler-options ["-Xlint:-options"]
        aliases          [:module.test/deps]
        verbose?         true
        compile?         true
        options          {:source-path      source-path
                          :target-path      target-path
                          :compiler-options compiler-options
                          :aliases          aliases
                          :verbose?         verbose?
                          :compile?         compile?}
        source-paths     (->> source-path
                           sut/find-java-paths
                           (map str)
                           set)
        includes?        #(string/includes? % "ninja/platform/schema/0.0.1-alpha1/schema-0.0.1-alpha1.jar")]

    (testing "expected failed compilation result"
      (let [bad-options (update options :source-path str "/bad-source-path")
            {:as res :keys [type data meta]} (sut/compile bad-options)]

        (testing "as a unified response"
          (is (r/error? res))
          (is (= :error type))
          (is (nil? meta)))

        (testing "with correct message"
          (is (= "Can't find the Java source files" (:message data))))

        (testing "with provided options without any changes"
          (is (= bad-options (:options data))))))


    (testing "expected successful compilation result"
      (let [{:as res :keys [type data meta]} (sut/compile options)
            compile-opts (:compile-opts data)
            command      (:command data)]

        (testing "as a unified response"
          (is (false? (r/error? res)))
          (is (= :success type))
          (is (nil? meta)))

        (testing "with correct message"
          (is (= "Processed 8 files" (:message data))))

        (testing "with provided options without any changes"
          (is (= options (:options data))))

        (testing "with zero compilation status"
          (is (= {:code 0, :info "", :warnings ""} (:compilation-result data))))

        (testing "with correct compile options"
          (is (= target-path (str (:target-path compile-opts))))
          (is (= compiler-options (:compiler-options compile-opts)))
          (is (every? #(contains? source-paths (str %)) (:source-paths compile-opts)))
          (is (true? (includes? (:classpath compile-opts)))))

        (testing "with correct command"
          (is (= 13 (count command)))
          (is (= "-cp" (first command)))
          (is (= (:classpath compile-opts) (second command)))
          (is (true? (includes? (second command))))
          (is (= compiler-options (take-while (partial not= "-d") (nnext command))))
          (is (= ["-d" target-path] (->> command nnext (drop-while (partial not= "-d")) (take 2))))
          (is (every? #(contains? source-paths %) (->> command nnext (drop-while (partial not= target-path)) next))))))

    ;; cleanup
    (sut/delete-dirs! target-path)))


javac/src/test/resources/fixtures/ninja/sandbox/ExampleEnum.java
package fixtures.ninja.sandbox;

public class ExampleEnum {
}


javac/src/test/resources/fixtures/ninja/sandbox/ExampleClass1.java
package fixtures.ninja.sandbox;

public class ExampleClass1 {
}


javac/src/main/clojure/ninja/javac.clj
(ns ninja.javac
  (:refer-clojure :exclude [compile])
  (:gen-class)
  (:require
    [clojure.edn :as edn]
    [clojure.java.io :as io]
    [clojure.string :as string]
    [clojure.tools.deps.alpha :as deps]
    [clojure.tools.deps.alpha.util.maven :as maven]
    [io.pedestal.log :as log]
    [ninja.response :as r])
  (:import
    (java.io
      ByteArrayOutputStream
      IOException)
    (java.nio.file
      FileSystemLoopException
      FileVisitOption
      FileVisitResult
      FileVisitor
      Files
      LinkOption
      NoSuchFileException
      Path
      Paths)
    (java.nio.file.attribute
      BasicFileAttributes
      FileAttribute)
    (java.util
      EnumSet)
    (javax.tools
      JavaCompiler
      ToolProvider)))


(set! *warn-on-reflection* true)


(defn ^Path make-path
  "Converts a path string, or a sequence of strings that when joined form a path string, to a Path.
  Returns an instance of `java.nio.file.Path`."
  [path & paths]
  (->> paths
    (map str)
    (into-array String)
    (Paths/get (str path))))


(defn make-classpath
  "Returns a calculated classpath taken from deps."
  ([]
    (make-classpath {}))

  ([{:keys [deps-map aliases]}]
    (let [deps-map (or deps-map (deps/slurp-deps (io/file "deps.edn")))
          deps-map (update deps-map :mvn/repos #(merge maven/standard-repos %))
          paths    (select-keys deps-map [:paths])
          args-map (deps/combine-aliases deps-map aliases)]
      (as-> deps-map $
        (deps/resolve-deps $ args-map)
        (deps/make-classpath-map paths $ args-map)
        (get $ :classpath-roots)
        (deps/join-classpath $)))))


(defn file-visitor
  "Accepts handlers and returns an instance of `java.nio.file.FileVisitor`."
  [{:keys [visit-file-fn visit-file-failed-fn pre-visit-directory-fn post-visit-directory-fn]
    :or   {pre-visit-directory-fn  (constantly FileVisitResult/CONTINUE)
           post-visit-directory-fn (constantly FileVisitResult/CONTINUE)
           visit-file-fn           (constantly FileVisitResult/CONTINUE)
           visit-file-failed-fn    (constantly FileVisitResult/SKIP_SUBTREE)}}]
  (reify FileVisitor
    (preVisitDirectory [_ dir attrs] (pre-visit-directory-fn dir attrs))
    (visitFile [_ file attrs] (visit-file-fn file attrs))
    (visitFileFailed [_ file e] (visit-file-failed-fn file e))
    (postVisitDirectory [_ dir e] (post-visit-directory-fn dir e))))


(defn as-path
  "Ensures an instance of `java.nio.file.Path`."
  [x]
  (if (instance? Path x) x (make-path x)))


(defn find-paths
  "Returns a collection of paths in the given directory filtered using the provided predicate."
  ([root]
    (find-paths root (constantly true)))

  ([root pred]
    (let [*paths (atom [])]
      (Files/walkFileTree (as-path root) (EnumSet/of FileVisitOption/FOLLOW_LINKS) Integer/MAX_VALUE
        (file-visitor
          {:visit-file-fn        (fn [^Path file ^BasicFileAttributes attrs]
                                   (when (pred file attrs)
                                     (swap! *paths conj file))
                                   FileVisitResult/CONTINUE)
           :visit-file-failed-fn (fn [_ ^IOException e]
                                   (cond
                                     (instance? FileSystemLoopException e) FileVisitResult/SKIP_SUBTREE
                                     (instance? NoSuchFileException e) FileVisitResult/SKIP_SUBTREE
                                     :else (throw e)))}))
      @*paths)))


(defn find-java-paths
  "Returns a collection of paths to the Java source files in the given directory."
  [root]
  (find-paths root
    (fn [^Path path ^BasicFileAttributes attrs]
      (and (.isRegularFile attrs)
        (string/ends-with? path ".java")))))


(defn exists?
  "Returns `true` if the file exists. Otherwise, `false`."
  ([path]
    (exists? path [LinkOption/NOFOLLOW_LINKS]))

  ([path link-opts]
    (Files/exists (as-path path) (into-array LinkOption link-opts))))


(defn create-dirs!
  "Creates directories using `java.nio.Files/createDirectories`. Also creates parents if needed."
  ([root]
    (create-dirs! root []))

  ([root attrs]
    (Files/createDirectories (as-path root) (into-array FileAttribute attrs))))


(defn delete-dirs!
  "Deletes the file tree. Doesn't follow symbolic links."
  [root]
  (Files/walkFileTree (as-path root) #{} Integer/MAX_VALUE
    (file-visitor
      {:visit-file-fn           (fn [^Path file _] (Files/delete file) FileVisitResult/CONTINUE)
       :post-visit-directory-fn (fn [^Path dir _] (Files/delete dir) FileVisitResult/CONTINUE)
       :visit-file-failed-fn    (fn [_ ^IOException e]
                                  (cond
                                    (instance? FileSystemLoopException e) FileVisitResult/SKIP_SUBTREE
                                    (instance? NoSuchFileException e) FileVisitResult/SKIP_SUBTREE
                                    :else (throw e)))})))


(defn make-command
  "Builds and returns a Java compiler command."
  [{:keys [classpath target-path compiler-options source-paths]
    :or   {target-path      "classes"
           compiler-options []
           source-paths     []}}]
  `["-cp" ~classpath
    ~@compiler-options
    "-d" ~(str target-path)
    ~@(map str source-paths)])


;; TODO: [2021-03-14, just.sultanov@gmail.com] Set `:verbose?` option to false (by default)
(defn compile
  "A simple wrapper for the Java source compiler.

  Params:
  * `:source-path`      - Path to java sources (Default: \"src/main/java\")
  * `:target-path`      - Files are compiled to the target path (Default: \"classes\")
  * `:compiler-options` - Java compiler options (Default: [])
  * `:deps-map`         - Deps configuration map. Reads `deps.edn` by default.
  * `:aliases`          - Additional aliases from `deps.edn` (Default: [])
  * `:verbose?`         - Print extra info (Default: true)
  * `:compile?`         - Is compilation enabled?. Can be useful for debugging. (Default: true)"
  [{:as   options
    :keys [source-path target-path compiler-options deps-map aliases verbose? compile?]
    :or   {source-path      "src/main/java"
           target-path      "classes"
           compiler-options []
           aliases          []
           verbose?         true
           compile?         true}}]
  (log/info :ninja.javac.compile/starting options)
  (let [compiler ^JavaCompiler (ToolProvider/getSystemJavaCompiler)]
    (if-not compiler
      (let [result {:message "Can't find the Java compiler"
                    :options options}]
        (log/error :ninja.javac.compile/failed result)
        (r/as-error result))
      (let [source-path  (make-path source-path)
            target-path  (make-path target-path)
            source-paths (find-java-paths source-path)]
        (if-not (seq source-paths)
          (let [result {:message "Can't find the Java source files"
                        :options options}]
            (log/error :ninja.javac.compile/failed result)
            (r/as-error result))
          (let [classpath    (make-classpath {:deps-map deps-map, :aliases aliases})
                compile-opts {:classpath        classpath
                              :target-path      target-path
                              :compiler-options compiler-options
                              :source-paths     source-paths}
                command      (make-command compile-opts)
                result       (cond-> {:message (format "Processed %s files" (count source-paths))
                                      :options options}
                               verbose? (assoc :command command :compile-opts compile-opts))]
            (if-not compile?
              (do
                (log/info :ninja.javac.compile/completed result)
                (r/as-success result))
              (let [_      (create-dirs! target-path)
                    out    (ByteArrayOutputStream.)
                    err    (ByteArrayOutputStream.)
                    code   (.run compiler nil out err (into-array String command))
                    result (assoc result :compilation-result {:code code, :info (str out), :warnings (str err)})]
                (log/info :ninja.javac.compile/completed result)
                (if (zero? code)
                  (r/as-success result)
                  (r/as-error (assoc result :message "Something went wrong")))))))))))


;; TODO: [2021-03-14, just.sultanov@gmail.com] Add CLI: provide all options and add help
(defn -main
  "The entry point to run compiler using `clojure -M`. Use the `compile` function directly."
  [& [source-path target-path deps-map aliases verbose? compile? & compiler-options]]
  (let [res (compile
              {:source-path      source-path
               :target-path      target-path
               :deps-map         (edn/read-string deps-map)
               :compiler-options compiler-options
               :aliases          (edn/read-string aliases)
               :verbose?         (edn/read-string verbose?)
               :compile?         (edn/read-string compile?)})]
    (when (r/error? res)
      (System/exit 1))))


(comment
  ;; example
  (compile {:source-path      "src/test/resources/fixtures"
            :target-path      "target/classes"
            :compiler-options ["-target" "15"
                               "-source" "15"
                               "-Xlint:-options"]
            :aliases          [:module.test/deps] ;; adds `ninja.platform/schema:0.0.1-alpha1` for testing classpath calculation
            :verbose?         false
            :compile?         true})

  ;; a similar command for the main entry point
  (-main
    "src/test/resources/fixtures" ;; :source-path
    "target/classes" ;; :target-path
    "nil" ;; :deps-map
    "[:module.test/deps]" ;; :aliases
    "true" ;; :verbose?
    "true" ;; :compile?
    "-target" "15" "-source" "15" "-Xlint:-options" ;; :compiler-options
    )
  )
